pkgs/_analyzer_macros/lib/analyzer_macro_implementation.dart

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:_macro_host/macro_host.dart';
+// ignore: implementation_imports
+import 'package:analyzer/src/summary2/macro_injected_impl.dart' as injected;
+import 'package:dart_model/dart_model.dart';
+import 'package:macro_service/macro_service.dart';
+import 'package:macros/macros.dart';
+// ignore: implementation_imports
+import 'package:macros/src/executor.dart' as injected;
+
+/// Injected macro implementation for the analyzer.
+class MacroImplementation implements injected.MacroImplementation, HostService {
+  final Uri packageConfig;
+  final Map<String, String> macroImplByName;
+  late final MacroHost _host;
+
+  MacroImplementation._(this.packageConfig, this.macroImplByName);
+
+  /// Starts the macro host.
+  ///
+  /// [packageConfig] is the package config of the workspace of the code being
+  /// edited.
+  ///
+  /// [macroImplByName] identifies macros, it's a placeholder until we identify
+  /// macros using package config. Keys are macro annotation qualified names
+  /// (`uri#name`) and values are macro implementation qualified names.
+  static Future<MacroImplementation> start({
+    required Uri packageConfig,
+    required Map<String, String> macroImplByName,
+  }) async {
+    final result = MacroImplementation._(packageConfig, macroImplByName);
+    result._host = await MacroHost.serve(
```

Comment:
I see this pattern is again repeated a bunch across these implementations. I will stop harping on it but something about the design here definitely seems off, that we require so much late initialization and cyclic dependencies.

Reply:
It's probably always correct to refactor, so please do keep mentioning :) thanks.

Here again it's due to the class doing two things; moved the host service to a separate class.